streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java

```diff
@@ -254,19 +261,22 @@ public void putAll(final List<KeyValue<Bytes, byte[]>> entries) {
 
     @Override
     public byte[] delete(final Bytes key) {
+        Objects.requireNonNull(key, "key cannot be null");
         validateStoreOpen();
         lock.writeLock().lock();
         try {
-            Objects.requireNonNull(key);
-            final byte[] v = getInternal(key);
-            cache.delete(cacheName, key);
-            underlying.delete(key);
-            return v;
+            return deleteInternal(key);
         } finally {
             lock.writeLock().unlock();
         }
     }
 
+    private byte[] deleteInternal(final Bytes key) {
```

Comment:
This is also intentional but I'd like to have at least a few more careful checks on: I think our current handling of delete, which is to remove from cache and remove from store immediately is not correct, since by doing this, the deletion will not be forwarded to downstream processors (even if the store is flushed later). So I changed it to just putting a null byte arrays into the cache, which will also be picked over the underlying store in the merge-sort iterators.

But I want to make sure that there are no other corner cases that may be affected by this change. @dguy @mjsax @bbejeck ?

Reply:
I believe this to be correct as well, as all the underlying stores handle the put of a `null` as a delete and it seems compatible with the new change to not have `all()` include null values.

Reply:
The reason I think we did not capture this issue earlier, is that `tuple forward in flushing` is only used for DSL, and in our current DSL's KTableXX operators we never explicitly trigger `delete` and hence that's why we never realize this issue.

But even without the scenario that could expose this bug, not using write-through caching for delete still have a performance benefits of write-back caching.